aten/src/ATen/cpu/vec/vec256/vec256_int.h

```diff
@@ -1133,6 +1133,117 @@ inline Vectorized<int8_t> Vectorized<int8_t>::le(const Vectorized<int8_t>& other
   return (*this <= other) & Vectorized<int8_t>(1);
 }
 
+template <>
+Vectorized<int64_t> inline operator<<(const Vectorized<int64_t>& a, const Vectorized<int64_t>& b) {
+  return _mm256_sllv_epi64(a, b);
+}
+
+template <>
+Vectorized<int32_t> inline operator<<(const Vectorized<int32_t>& a, const Vectorized<int32_t>& b) {
+  return _mm256_sllv_epi32(a, b);
+}
+
+template <>
+Vectorized<int16_t> inline operator<<(const Vectorized<int16_t>& a, const Vectorized<int16_t>& b) {
+  // No vector instruction for shifting int16_t, so emulating it instead.
+
+  // Mask used to set upper 16 bits of each 32-bit value to 0, and
+  // keep lower 16 bits.
+  __m256i mask = _mm256_set1_epi32(0xFFFF);
+
+  // Convert 16-bit operands from lane #0 to 32-bit values, and
+  // perform vectorized shift.  Make sure that upper 24 bits of 32-bit
+  // results are all 0.
+  __m128i a_lo_16 = _mm256_extracti128_si256(a, 0);
+  __m128i b_lo_16 = _mm256_extracti128_si256(b, 0);
+  __m256i a_lo_32 = _mm256_cvtepi16_epi32(a_lo_16);
+  __m256i b_lo_32 = _mm256_cvtepi16_epi32(b_lo_16);
+  __m256i c_lo_32 = _mm256_and_si256(_mm256_sllv_epi32(a_lo_32, b_lo_32), mask);
```

Comment:
This operation is probably memory bound anyway, but I don't think you need to cross AVX lanes here. If you shuffle neighboring pairs of int16 from `a b c d ...` into two vectors `0 a 0 c ...` and `0 b 0 d ...` then shifting  should give the same answer as if you had used `cvtepi16_epi32` but without needing to cross AVX lanes.

Reply:
Thanks, I've made the change and it's indeed faster when crossing lanes avoided.  Will change the other two functions too, and update PR then.